Sources/Amplitude/Amplitude.swift

```
@@ -8,7 +8,70 @@ public class Amplitude {
         sessions.sessionId
     }
 
-    var state: State = State()
+    private let identityLock = NSLock()
+    private var _identity = Identity()
+    var identity: Identity {
+        get {
+            identityLock.withLock {
+                return _identity
+            }
+        }
+        set {
+            applyIdentityUpdate(newValue)
+        }
+    }
+
+    private func applyIdentityUpdate(_ identity: Identity, sendIdentifyIfNeeded: Bool = true) {
+        var deviceIdChanged = false
+        var userIdChanged = false
+        var userPropertiesChanged = false
+        identityLock.withLock {
+            let oldValue = _identity
+            _identity = identity
+
+            if identity.deviceId != oldValue.deviceId {
+                deviceIdChanged = true
+                try? storage.write(key: .DEVICE_ID, value: identity.deviceId)
```

`applyIdentityUpdate` may be called on the main thread. In PersistentStorage’s write implementation, the task is attached to the queue using `sync`. This might cause some thread blocking. How about changing the write operation in PersistentStorage to `async`?

---

This currently maintains the existing behavior, and I agree this would be better async. But, it means sdk state and storage state would be out of sync, and I'm not sure if we have anything that would break with that behavior. Something like `Amplitude(config).setUserId(""); Amplitude(config).getUserId()` would return different values, though this is probably is safe to ignore. Anything else?  We're still ignoring errors on write, so this isn't actually that different.

It looks like writes are already ordered with a sync storage queue so we could just reuse that and maintain write ordering while being async- should a quick change.

---

Are you referring to different instances for the userId example? In this case, there might indeed be issues, but they can be ignored. Read operations within the same instance are either during the initialization phase or independent file transfer operations, so there shouldn’t be any problems.

---

I am referring to initializing the same instance twice, which will no longer work.